plugins/kubernetes-backend/src/service/KubernetesErrorIndicator.ts

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Logger } from 'winston';
+import {
+  KubernetesFetchError,
+  KubernetesErrorTypes,
+} from '@backstage/plugin-kubernetes-common';
+import { KubernetesRejectionHandler } from '../types/types';
+import { HttpError } from '@kubernetes/client-node';
+
+const knownStatusCodeErrors: Record<number, KubernetesErrorTypes> = {
+  400: 'BAD_REQUEST',
+  401: 'UNAUTHORIZED_ERROR',
+  404: 'NOT_FOUND',
+  500: 'SYSTEM_ERROR',
+};
+
+export class KubernetesErrorIndicator implements KubernetesRejectionHandler {
+  private readonly logger: Logger;
+  private readonly statusCodeErrorTypes: Record<number, KubernetesErrorTypes>;
+  private readonly signature: string = 'KubernetesFetchError';
+  constructor(
+    logger: Logger,
+    statusCodeErrorTypes: Record<
+      number,
+      KubernetesErrorTypes
+    > = knownStatusCodeErrors,
+  ) {
+    this.logger = logger;
+    this.statusCodeErrorTypes = statusCodeErrorTypes;
+  }
+  public onRejected: (
+    reason: any,
+    resourcePath?: string,
+  ) => Promise<KubernetesFetchError> = (reason, resourcePath) => {
+    if (reason.name === 'HttpError' && reason.statusCode) {
+      const { body } = reason as HttpError;
+      this.logger.warn(
+        `statusCode=${
+          reason.statusCode
+        } for resource ${resourcePath} body=[${JSON.stringify(body)}]`,
+      );
+      return Promise.resolve({
+        errorType:
+          this.statusCodeErrorTypes[reason.statusCode] || 'UNKNOWN_ERROR',
+        statusCode: reason.statusCode,
+      });
+    }
+    if (this.didIntercept(reason)) {
```

Comment:
I am not sure I understand this. Are you ensuring that there's a rejection the first time it's seen but a resolution the second time on? This seems like it's a bit convoluted and potentially hard to follow. Maybe I'm misunderstanding :thinking:

Reply:
There's no more indicator and throw-and-catch stuff in the new PR!